server/datastore/mysql/software.go

```diff
@@ -2877,301 +2838,536 @@ last_vpp_install AS (
+			args...,
+		); err != nil {
+			return nil, nil, ctxerr.Wrap(ctx, err, "get host software count")
+		}
+
+		var replacements []any
+		if len(softwareTitleIds) > 0 {
+			replacements = append(replacements,
+				// For software installers
+				`
+				SELECT
+					software_titles.id,
+					software_titles.name,
+					software_titles.source AS source,
+					software_installers.self_service AS package_self_service,
+					software_installers.filename AS package_name,
+					software_installers.version AS package_version,
+					software.id AS software_id,
+					software.source AS software_source,
+					software.version AS version,
+					software.bundle_identifier AS bundle_identifier,
+					NULL AS vpp_app_self_service,
+					NULL AS vpp_app_adam_id,
+					NULL AS vpp_app_version,
+					NULL AS vpp_app_platform,
+					NULL AS vpp_app_icon_url,
+					NULL AS last_install_installed_at,
+					NULL AS last_install_install_uuid,
+					NULL AS last_uninstall_uninstalled_at,
+					NULL AS last_uninstall_script_execution_id,
+					GROUP_CONCAT(software_cve.cve) AS vulnerabilities_list
+			`, cveMetaFilter, `
+				GROUP BY
+					software_titles.id,
+					software_titles.name,
+					software_titles.source,
+					software_installers.self_service,
+					software_installers.filename,
+					software_installers.version,
+					software.id,
+					software.source,
+					software.version,
+					software.bundle_identifier
+			`)
+		}
+		if len(vppAdamIDs) > 0 {
+			replacements = append(replacements,
+				// For vpp apps
+				`
+				SELECT
+					software_titles.id,
+					software_titles.name,
+					software_titles.source AS source,
+					software_installers.self_service AS package_self_service,
+					software_installers.filename AS package_name,
+					software_installers.version AS package_version,
+					NULL AS software_id,
+					NULL AS software_source,
+					NULL AS version,
+					NULL AS bundle_identifier,
+					vpp_apps_teams.self_service AS vpp_app_self_service,
+					vpp_apps_teams.adam_id AS vpp_app_adam_id,
+					vpp_apps.latest_version AS vpp_app_version,
+					vpp_apps.platform as vpp_app_platform,
+					NULLIF(vpp_apps.icon_url, '') AS vpp_app_icon_url,
+					NULL AS last_install_installed_at,
+					NULL As last_install_install_uuid,
+					NULL AS last_uninstall_uninstalled_at,
+					NULL AS last_uninstall_script_execution_id,
+					GROUP_CONCAT(software_cve.cve) AS vulnerabilities_list
+			`, cveMetaFilter, `
+				GROUP BY
+					software_titles.id,
+					software_titles.name,
+					software_titles.source,
+					software_installers.self_service,
+					software_installers.filename,
+					software_installers.version,
+					vpp_apps_teams.self_service,
+					vpp_apps_teams.adam_id,
+					vpp_apps.latest_version,
+					vpp_apps.platform,
+					vpp_apps.icon_url
+			`)
+		}
+		stmt = fmt.Sprintf(stmt, replacements...)
+		stmt = fmt.Sprintf("SELECT * FROM (%s) AS combined_results", stmt)
+		// stmt, _ = appendListOptionsToSQL(stmt, &opts.ListOptions)
```

Comment:
Right now we return all of the hosts results and paginate in memory. This need not be the case. We should be able to adjust the `group by` and return only software titles. The would impact the versions and vulnerability list. However, not unlike in the original code, we could pull up the versions and vulnerabilities as separate queries and results merged in.